### ISSUE TITLE
[FEA] Consider disabling `--expt-relaxed-constexpr`

### ISSUE THREAD

**Issue by jrhemstad** [rapidsai/cudf#7795]
**Is your feature request related to a problem? Please describe.**

`--expt-relaxed-constexpr` is a convenient way to reuse existing `constexpr` host code, e.g., things like `std::max`.

However, it can lead to some pretty surprising behavior. Consider:

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**
Add any other context, code examples, or references to existing implementations about the feature request here.

**Comment by harrism** [comment]
@karthikeyann please be aware of this with respect to #7713

**Comment by harrism** [comment]
> At the very least, we should be preferring CUDA_HOST_DEVICE_CALLABLE whenever possible.

No, we should *prefer* `CUDA_DEVICE_CALLABLE` or no annotation whenever possible. We should use `CUDA_HOST_DEVICE_CALLABLE` only for functions that are required to be called from both host and device.

**Comment by jrhemstad** [comment]
> only for functions that are required to be called from both host and device.

This was implied. The whole point of `--expt-relaxed-constexpr` is to call existing host `constexpr` functions from device, which implies it can be/is called from both host and device.

**Comment by harrism** [comment]
I want to be very explicit because it is not obvious to everyone (a fact that is evident in our code).

**Comment by harrism** [comment]
Still valid

**Comment by PointKernel** [comment]
This seems to be unblocked with the recent CCCL update.

The changes in [NVIDIA cuCollections PR #595](https://github.com/NVIDIA/cuCollections/pull/595) show that cuco can now build and run properly without relaxed constexprs.

**Comment by vyasr** [comment]
nvidia/cucollections#595 and rapidsai/cuspatial#1494 inspired me to look into this again. It's slow going (but low effort) to investigate, so I'll report back when I get compilation far enough to make a determination on whether removing this setting will be possible in cudf. I've run into one very odd case that seemed like a compiler bug, but other than that it seems largely like a lot of tedious but not difficult work.

**Comment by vyasr** [comment]
@PointKernel any idea where we are with this now? Is https://github.com/rapidsai/cudf/issues/17797 the last major blocker or are there other issues?

**Comment by PointKernel** [comment]
> [@PointKernel](https://github.com/PointKernel) any idea where we are with this now? Is [#17797](https://github.com/rapidsai/cudf/issues/17797) the last major blocker or are there other issues?

You're right, #17797 has been the main blocker, and we were waiting on CCCL updates to move forward with other parts of libcudf. Now that CCCL 3.0 is in, I'll go ahead and reprioritize this.

**Comment by vyasr** [comment]
Awesome sounds good!

**Comment by PointKernel** [comment]
For future reference: I previously discussed this issue with Vyas offline and outlined a list of tasks to address it, but I never shared the outcome here. One task was to move `CUDF_HOST_DEVICE` APIs from `.hpp` headers to `.cuh` headers, since they contain device code. However, as David pointed out https://github.com/rapidsai/cudf/pull/19541#discussion_r2270621445, we actually want these APIs to remain in `.hpp` headers; otherwise, `.cpp` files could end up including `.cuh` headers unnecessarily, even when no device functions are used.
